### DISCUSSION TITLE
Trackable async handlers or other suggestions?

### DISCUSSION THREAD

**Discussion by domehead100** [mobxjs/mobx#3099] [Ideas]
I've been using Mobx for about 4-5 years now on a large project, Mobx has been a critical piece in helping us create a superlative developer experience wherein creating new features is easy, changing things is easy, and fixing UI defects is trivial.

One recurring problem is when developers make an api call in the onClick handler of a button and don't disable the button while the api call is pending, which can result in an n-clickable button performing the same api call n times.

Having attempted to tackle this in various ways that I won't go into right now, an additional thought was, at least in the case of a button, to make the click handler function itself have an observable isPending property.  Then a button component might just look like:

```jsx
<Button
    ...
   onClick={someHandlerMethodThatMakesAnApiCall}
   disabled={!hasValidationErrors || someHandlerMethodThatMakesAnApiCall.isPending}
/>
```

Just now while experimenting with this, I quickly hacked out a small function...

```typescript
export function makeTrackable<TReturnType = void>(fn: (...args) => TReturnType) {
    if (typeof fn === "function") {
        (fn as any).isPending = observable.box(false);
    }
    return fn as ((...args) => TReturnType) & { isPending: IObservableValue<boolean> };
}
```
... that takes a handler and adds ths isPending property to the function itself (functions are objects after all), and that isPending property can be tracked to make it easier to disable the button while the api call made via the handler is in progress.  

My handler then becomes ```makeTrackable(<original handler>);```, and the developer still has to do isPending.set(true) near the top of the handler before making the api call and isPending.set(false) in a .finally clause of the api promise, but at least it's logical and easy and requires no additional state variables cluttering up the store(s).

Some issues though:
1) If the handler is a Mobx action, the isPending property is not surfaced since action wraps the original function in a new function and while adding the property .isMobxAction to the wrapper function, it does not copy any properties from the original function to the wrapper function.  So, the above does not work at all if the handler is a Mobx action.  Also, even if an action did surface our pending property, it still would not trigger state changes until the transaction is complete, so this wouldn't work for us either.  It would be a simple change for action to reference any properties on the original function, and I can't remember: is there some way to trigger reactions inside of an action?

2) I'd prefer to use extendObservable to add the pending property to the handler function, but I can't do this because in __DEV__ mode extendObservable checks that typeof target === "object", and since a function returns a type of "function" (even though a function is an object), extendObservable throws an error in __DEV__ mode instead of adding the observable property.  In production builds of Mobx, using extendObservable works fine.  It would be a minor convenience for a developer to be able to use assignment instead of .set() and for the disabled prop on the button to just reference .isPending instead of isPending.get().

Anyway, just interested in suggestions, ideas, or feedback.

**Comment by urugator** [comment]
> Pending property is not surfaced since action wraps the original function

Can you post an example of how do you intend to use `makeTrackable`? Can't you wrap the function with `action` inside `makeTrackable` and define the prop on the action? I think that `observable/makeObservable/etc` won't attempt to turn a prop into an action if it's an action already...
Also it's not clear to me, how do you access `isPending` from within the function.

>  it still would not trigger state changes until the transaction is complete

Assuming the "n-click" scenario, afaik that's not an issue. Everything happens synchronously, react will re-render with updated `isPending`, before the second click can be noticed by browser.

> add the pending property to the handler functio
```typescript
export function makeTrackable<TReturnType = void>(fn: (...args) => TReturnType) {
    if (typeof fn === "function") {
        const box = observable.box(false); 
        Object.defineProperty(fn, 'isPending', {		
          get() { box.get() },
          set(value) { box.set(value) }
        })   
    }
    return fn as ((...args) => TReturnType) & { isPending: IObservableValue<boolean> };
}
```

**Comment by mweststrate** [comment]
Only scanned the issue, but looks like
https://github.com/mobxjs/mobx-utils#frompromise might be applicable here?

**Comment by domehead100** [comment]
Thanks Michel.  We use fromPromise extensively (as in 99% of our async calls are using it), and it was my first thought too, but it doesn’t apply here because I need to know if the function is pending without executing the function to get a result to find that out. 

 I’m just trying to compress/reduce state clutter here so to speak by storing the pending state of the function on the function itself, which I can then use, for example, to disable buttons while their click handlers are executing.

An example should help clarify.  I’ll stick one in next post.

**Comment by domehead100** [comment]
Thanks @urugator, I was kind of hoping you'd chime in :).

So for now, here's an example using the original makeTrackable that I posted initially

```typescript
// store.ts
// no @action because it will hide saveBulkChanges and isPending inside of the action wrapper function closure
saveBulkChanges = makeTrackable(() => {
    const { doSomething } = this;
    this.saveBulkChanges.isPending.set(true);
    Api.post("/some-route").then(data => {
        doSomething(data);
    }).finally(() => {
        this.saveBulkChanes.isPending.set(false);
    }
}
```
```typescript
// component.tsx
<Button onClick={saveBulkChanges} disabled={saveBulkChanges.isPending.get()}>Save changes</Button>
```

Now this does require the developer to use makeTrackable and it does require them to remember to set the isPending value before and after the async/api call is made, but its still an improvement in my opinion over having a separate state variable for tracking each pending call cluttering up state, a single state variable map that can be used to track any pending call, or having the Api keep track of the pending calls in its own map where they can be referenced.  All of these approaches work fine, but they suffer either from lack of locality to the function itself (affecting the developer's mental model) or cluttering of state or for some other reason not being used by lazy or forgetful developers :0;

Thanks for pointing out that the action transaction is not a problem because the setting of isPending to true happens synchronously before render and then the setting it to false happens later after the api call's promise has resolved.

Also, neat trick on the getter/setter to allow direct assignment or referencing the boxed observable, but I think the .get()/.set() on the observable is actually okay (despite what I might have said originally) compared to keeping the observable in a closure outside of the function.  I was trying to use extendObservable where you are using a closure-scoped variable and Object.defineProperty.  If extendObservable would change its check for ```typeof target === "object"``` to ```typeof target === "object" || typeof target === "function"```, then extendObservable would work.  Well, as mentioned it actually does work, just not in a dev build where that check throws an error instead.

Interesting idea on having makeTrackable create the action and then putting isPending onto the action wrapper function.

So, taking your idea of putting the isPending into closure scope, I hacked together the method below.  In this case, the developer does not have to manually set the isPending.  Rather, it is set to true before the original function is called and then if the original function returns a promise then isPending is set to false after that promise resolves (via finally) or immediately if the function does not return a promise.  Having tried this out, it seems to work.

```typescript
export function makeAutoTrackable<TReturnType = void>(fn: (...args) => TReturnType) {
    let _fn = fn;
    if (typeof fn === "function") {
        const isPending = observable.box(false);
        _fn = (...args) => {
            isPending.set(true);
            // eslint-disable-next-line prefer-rest-params
            const rv: TReturnType = fn.apply(this, args);
            if ((rv as any).finally) {
                (rv as any).finally(() => {
                    isPending.set(false);
                });
            } else isPending.set(false);
            return rv;
        };
        (_fn as any).isPending = isPending;
    }
    return _fn as ((...args) => TReturnType) & { isPending: IObservableValue<boolean> };
}
```

And then this is just the same thing but the action is created inside the makeAutoTrackableAction function:

```typescript
export function makeAutoTrackableAction<TReturnType = void>(fn: (...args) => TReturnType) {
    let _fn = fn;
    if (typeof fn === "function") {
        const isPending = observable.box(false);
        _fn = action((...args) => {
            isPending.set(true);
            // eslint-disable-next-line prefer-rest-params
            const rv: TReturnType = fn.apply(this, args);
            if ((rv as any).finally) {
                (rv as any).finally(() => {
                    isPending.set(false);
                });
            } else isPending.set(false);
            return rv;
        });
        (_fn as any).isPending = isPending;
    }
    return _fn as ((...args) => TReturnType) & { isPending: IObservableValue<boolean> };
}
```

I'm thinking that I'll be adding four makeTrackable functions above to our code base and try them out: one where isPending must be manually set and not an action, one where isPending must be manually set and it's an action, and the two "auto" flavors above where the developer can return a promise and the isPending will be automatically set.  That should give us enough options to meet the various use cases and developer preferences.

**Comment by urugator** [comment]
> If extendObservable would change its check for typeof target === "object" to typeof target === "object" || typeof target === "function"

Most likely not gonna happen. Function has extra props like `name`, `length`,`arguments` that we would probably require extra care.
We would probably have to adapt `make(Auto)Observable` as well. There are no tests for this, so someone would have to write them.
Imo it's an edge case and reasonable workaround exists. 

> neat trick on the getter/setter to allow direct assignment or referencing the boxed observable

`extend/make/observable` impl does almost the same thing, but the suggested workaround will perform a lot better.

> compared to keeping the observable in a closure outside of the function

Can also be part of the function instead of a closure:
```javascript
fn._isPendingBox = observable.box(false);
Object.defineProperty(fn, 'isPending', {		
  get() { this._isPendingBox.get() },
  set(value) { this._isPendingBox.set(value) }
}) 
```

> one where isPending must be manually set

What I don't like about this the most, is that it's really not self-contained. The function must be defined on some object and it's implementation depends on the name of the property it's assigned to.
If I were to expose the `isPending` to the function, I would probably do it somehow like this:
```javascript
saveBulkChanges = makeTrackable(ctx => (arg1, arg2) => { 
  ctx.isPending = true;
  Api.post("/some-route").then(data => {
        doSomething(data);
  }).finally(() => {
    ctx.isPending = false;
  }
})
```
